### ISSUE TITLE
Make it easier to create connectors for integrating custom components

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#4925]
When integrating a component that has non-trivial functionality, it is usually necessary to use some custom JavaScript logic to connect things together. Implementing connector logic using `Element::executeJavaScript` becomes impractical if there are many JS snippets or if any of them are more complex than only a couple of statements. To deal with that, the connector logic can be extracted to a separate JavaScript file that exports functions that can be called using `executeJavaScript`.

Developing this kind of connector script could be further simplified by introducing some helper API around the concept.

One good way of structuring the JavaScript logic is as an ES6 class to which the target element instance is passed as a constructor parameter. To avoid naming conflicts, it would be recommended to name the class by converting the targeted custom element name to camel case and appending `Connector`, e.g. `VaadinTextFieldConnector`. 
```
class VaadinTextFieldConnector {
  constructor(element) {
    this.element = element;
  }
  setRequiredIndicatorVisible(visible) {
    // ...
  }
}
```

To access the logic defined by the ES6 class, a corresponding Java interface is defined. By default, the simple name of the interface should match the ES6 class name, but a custom ES6 class name can also be configured using an annotation on the interface. The interface should also define resource annotations (typically `@JavaScript`) for the files(s) containing the client-side implementation. It might be necessary to supplement the logic that collects resources for the production bundle to also take connector classes into account.

```
@JavaScript("vaadinTextFieldConnector.js");
public interface VaadinTextFieldConnector extends Connector {
  void setRequiredIndicatorVisible(boolean visible);
}
```

The framework can optionally parse the contents of the JavaScript file and report warnings if the ES6 class doesn't contain methods corresponding the the methods defined in the Java interface.

To use a connector, the component implementation uses the `public <T extends Connector> T createConnector(Class<T> type)` method defined in `Element`. This creates a new proxy instance and enqueues commands to create a corresponding client-side instance connected to the target element. The same commands will automatically be sent again if the element has been detached and becomes attached again.

Calling any method on the proxy instance will enqueue a command to call the corresponding JavaScript method. All the same parameter types as for `executeJavaScript` are supported. If `executeJavaScript` is enhanced to support getting a return value back from the client, then the connector proxy logic should also be updated to use the same mechanism. Until then, all methods in the interface must have `void` as their return type.

```
private final VaadinTextFieldConnector connector = getElement().createConnector(VaadinTextFieldConnector.class);

public void setRequiredIndicatorVisible(boolean visible) {
  super.setRequiredIndicatorVisible(visible);
  conenctor.setRequiredIndicatorVisible(visible);
}
```

In some cases, the client-side logic also needs to send messages to the server. To support this case, there is a two-arguments overload of `createConnector` that also receives an instance that serves as a reverse RPC target, similar to `ServerRpc` in Vaadin 7 and 8. A JavaScript representation of this object will be passed as a second parameter to the ES6 constructor. Calling methods on that object will trigger a request to the server that will invoke the corresponding method on the provided Java instance.

Example interface
```
public interface GridServerRpc extends ServerRpc {
  void select(String key);
  void deselect(String key);
}
```

Usage in the component
```
public Grid() {
  this.connector = getElement().createConnector(VaadinGridConnector.class, new GridServerRpc() {
    @Override
    public void select(String key) {
      getSelectionModel().selectFromClient(findByKey(key));
    }
    @Override
    public void deselect(String key) {
      getSelectionModel().deselectFromClient(findByKey(key));
    }
  });
}
```

Usage in JavaScript
```
class VaadinGridConnector {
  constructor(element, rpc) {
    // Assuming a super-convenient API in the element just to keep this example simple
    element.addSelectListener(event => rpc.select(event.key));
    element.addDeselectListener(event => rpc.deselect(event.key));
  }
}
```

**Comment by Legioth** [comment]
The `ServerRpc` mechanism is quite similar to the `$server` functionality that is available for any element based on `@ClientCallable` methods on a component class.

Similarities
* Client-side representation in the form of an object containing methods that trigger RPC
* Parameter deserialization semantics
* Targeting methods on a server-side object instance

Difference
* The JS object is serialized in slightly different ways - attached as an element propery vs passed to a constructor
* The methods to export are collected in slightly different ways - filtered by an annotation vs all public methods (except the ones inherited from `Object`).
* The server-side instance is slightly different - shared for other purpose vs created only for this purpose.

It seems relatively simple to create a slightly more abstract common implementation that is shared between both cases. Publishing can be defined as a JS expression to run in some context (or allow passing a representation through `executeJavaScript`). Registration can be based on instance + method filter.

**Comment by Legioth** [comment]
It might also be useful to have a variant that is directly mapped to the element instance on the client instead of an instance of a class provided by the integration. This instance would thus "only" be a typesafe way of doing `Element.callFunction`, and it could furthermore also be extended to allow things such as setting element properties or adding event listeners in a more typesafe way.

The real power would come from changing the API generator concept to generate this kind of internal connector instead of the current approach where the component API itself is generated. The benefit there is that the API generation can have more rough edges when the generated API is not directly exposed through the Component API but instead produces a class that is only used internally by the actual component API.

**Comment by Legioth** [comment]
To clarify how this feature would be used, we can use the integration in https://www.flowingcode.com/2019/01/creating-vaadin-flow-server-side-api.html as an example.

Instead of concatenating a string to build an init function that is passed to `executeJavaScript`, there would be a `/src/main/webapp/frontend/visConnector.js` file that contains the corresponding logic.
```
class VisConnector {
  constructor(element) {
    this.element = element;
  }
  init(serverItems) {
    var items = new vis.DataSet(serverItems);
    var options = {};
    this.timeline = new vis.Timeline(this.element, items, options);
  }
}
```

There would be a corresponding server-side connector interface:
```
public interface VisConnector {
  void init(Item[] serverItems);
}
```

The `Timeline` class would get an additional `@JavaScript("visConnector.js")` annotation so that the file containing the client-side connector class would be loaded. A connector instance can be created e.g. as an instance field with an initializer:
```
private final VisConnector connector = getElement().createConnector(VisConnector.class);
```
Finally, the component's constructor would pass the items to the the `init` method in the connector:
```
public Timeline(Item ... items) {
  connector.init(items);
}
```

To add support for focusing an item, you wold add a method for that in the server-side connector interface
```
void focus(int itemId);
```
implement the corresponding method in the client-side class
```
focus(itemId) {
  this.timeline.focus(itemId);
}
```
and finally implement the server-side component method to delegate to the connector
```
public void focus(Item item) {
 connector.focus(item.getId());
}
```

Finally, to show how how to use the reverse RPC feature, the item click handling could be implemented using that feature as an alternative to listening to the DOM event from the server.

The first step would be to pass a callback handler when creating the server-side connector instance:
```
private final VisConnector connector = getElement().createConnector(VisConnector.class, new AbstractConnectorCallback() {
  public void itemClicked(int itemId) {
    fireEvent(new ItemClickEvent(Timeline.this, true, itemId);
  }
});
```
The connector constructor would receive an RPC proxy as the second argument to the constructor. In this case, the DOM listener can be registered immediately in the constructor as well.
```
constructor(element, rpc) {
  this.element = element;
  this.element.addEventListener("click", event => {
    if (this.timeline) {
      rpc.itemClicked(this.timeline.getEventProperties(event).item);
    }
  });
}
```

**Comment by Legioth** [comment]
The API for defining an RPC handler is slightly challenging. My initial idea with a separate callback interface extending `ServerRpc` combined with an anonymous instance when creating the connector seems quite straightforward, but requires some boilerplate from defining a separate interface that doesn't really serve no purpose. The previous example skipped the interface by directly creating an anonymous instance of some base type. This is slightly weird because it's not conventional for an anonymous class to declare its own public methods.

A third alternative would be to export existing methods from the component class, similar to `@ClientCallable`. This creates a small challenge since the rest of the API is based on `Element` instead of `Component` and also be because there can in theory be multiple different connectors for the same element instance, and in the extreme case even multiple connectors of the same type. The problem with `Element` vs `Component` could be solved by still also offering a low-level API without ties to `Component`.  The problem with multiple connector types could be solved by declaring the target connector type in the annotation that is used to mark instance methods that should be used for the connector. Finally the edge case with multiple connector instances of the same type could be solved by allowing a connector instance as an additional parameter to a method. This parameter would thus be populated on the server and not be present in the client-side invokation.

With this kind of API, the previous example with clicks for the timeline component could instead be implemented with passing the component instance as the second parameter when registering the connector
```
private final VisConnector connector = getElement().createConnector(VisConnector.class, this);
```
and then declaring an annotated instance method that should be exported
```
@ConnectorCallbable(VisConnector.class)
private void itemClicked(int itemId) {
  fireEvent(new ItemClickEvent(this, true, itemId));
}
```
Client-side usage would be identical to the previous example.

**Comment by pleku** [comment]
This would be a significant improvement over the current approach, but I wonder if it could be streamlined a bit. I mean that there are quite many things to do for e.g. something as simple as the `focus(itemId)` method on the client side: 1) client side method, 2) connector method, 3) public API method.

I've head some feedback from a similar case: requiring the template model for simple types requires quite a lot of boilerplate, and would be much nicer to avoid that by being able to declare the API only once - meaning not having to duplicate it in the interface and in the component all the time for passing simple types from the bean to the client side. And even more complex types could be passed without the interface 

My initial thought was to make the component implement the interface. This means that the component or the interface must have implementations for the methods. Then we cannot allow different types to be declared in the public API and in the client API. Which lead me to think we should have some method that interacts with the client side (that and also we should not make the component a proxy instance).
```
VisConnector extends ClienConnector { // client connector could extend HasElement
   default void focus(Item item) {
      getElement().getConnector().callMethod("focus", item.getId());
   }
}
```
Then the component would just declare the public API by `implements VisConnector`. But this leads to the connector interface to be actually mostly unnecessary, we could just allow calling connector methods directly from the component methods:
```
public void focus(Item item) {
    getElement().getConnector().callMethod("focus", item.getId());
}
```

The instance returned by `getConnector()` would just be adaptor for e.g. `element.connector` property with some special semantics that map the client side connector's functions accordingly. 

The only advantage of having the connector interface would be to make it easier to discover the methods that interact with the client side connector, but I don't think that is required.


---
Not related to above, but I don't understand why `this` is required in 
```
private final VisConnector connector = getElement().createConnector(VisConnector.class, this);
```
since the component instance should be available for the `Element` already ?

**Comment by Legioth** [comment]
I like the idea of a generic dispatch method that accepts a method name and an arbitrary list of parameters. I would name it `callFunction` instead of `callMethod` to align with the very similar method in `Element`.  This approach would significantly simplify the implementation since there would be no need to mess with dynamically created proxy instances.

Support for defining a typesafe interface could be added at a later point if it's deemed necessary. Another thing of value that it does add is the possibility to automatically check that the method names in the interface are also present in the JavaScript file and that the number or parameters matches. Taken one step further, the warning message from the checker could contain a method stub for anything that is missing from the interface. In that way, the developer could just copy the basic structure into their Java code and fill in the actual type information manually. Full automatic generation would be problematic since the parameter types cannot be inferred from the JavaScript.

`getElement().getConnector()` is problematic since it assumes that there's only one connector instance per element. This is not the case when the super type defines one connector and a subclass defines another. Furthermore, it leaves no (good) way of defining the name of the client-side class.

`this` is used in the `createConnector` example since there would be multiple components associated with one element when using `Composite`.

**Comment by pleku** [comment]
>Full automatic generation would be problematic since the parameter types cannot be inferred from the JavaScript.

We could utilize _the component generator_, which should be able to quite easily generate the connector interfaces based Jsdocs. But getting around to that is miles away and would actually require people to create the Jsdocs and keep those up to date, which might cause more overhead than added value (in short term at least).

**Comment by Legioth** [comment]
The component generator approach works nicely when there are already JSDocs for other reasons. 

Maintaining JSDocs in your own connector JS just for the sake of generating a Java interface feels like more effort than just writing that interface by hand in the first place.

**Comment by Legioth** [comment]
There's one problem with the approach that would be based on a generic `callFunction` method: type support for parameters.

Current similar methods such as `Page.executeJavaScript` and `Element.callFunction` are on purpose limited to a small set of supported parameter types (which all have to implement `Serializable`). Passing beans or e.g. collections faces the same challenges as described in https://github.com/vaadin/flow/issues/5404.

This leaves us with one of these options:
* Stick to the limited type support, which means that my original example that passes `List<Item> serverItems` to the `init` function would have to be made more complex. We can later on add another abstraction that would automatically deal with more complex types.
* Define the parameters of `Connector.callFunction` as `String functionName, Object... parameters` and implement custom logic for dealing with more complex types. This would mean that `Connector.callFunction` would not be consistent with `Element.callFunction`.
* Update the existing `callFunction` and `executeJavaScript` methods to support a wider range of types. This would be a binary incompatible change which means that we should preferably do it at the same time when we're introducing the new `callJsFunction` and `executeJs` methods for dealing with binary incompatibility of the return type.

**Comment by pleku** [comment]
We can deliver quite much value already with the basic types support.
Adding array, collections and beans support can be added in later increments, and those are really complex to implement with an API that works always, we can also deliver value with examples and demos on how to solve the problems - it doesn't have to go into features.

**Comment by Legioth** [comment]
The main challenge with adding later is that we might at that point in time  again face problems with binary incompatible API changes.

**Comment by Legioth** [comment]
This idea does also need some tweaking in light of the ideas needed for ensuring CSP compatibility and improving type safety that are described in https://github.com/vaadin/flow/issues/10759. On the other hand, the proposed syntax of `someElement.getJsInvoker(GreeterJs.class).showGreeting("Hello");` does also satisfy the requirements we have here, namely a way of defining client-side code in a standalone file and then invoking that code in a way that is associated with a target element. The main difference is that there would be no separate "connector" instance but the exported functions would instead be invoked with `this` referencing the target DOM element. A connector implementation that needs its own state could thus hook on to the element instance (preferably by using a `WeakMap`, but also possible to just add its own state as a property on the element).

The original example with the required indicator could thus be implemented with code like this:
```
setRequiredIndicatorVisible(visible) {
  // ...
}
```

**Comment by Legioth** [comment]
Closing in favor of https://github.com/vaadin/flow/issues/10759. We might need some additional helpers for state management on top of the concept described in that ticket but let's worry about that when it becomes relevant.
